NPE when using Inline Refactoring on method with varargs within switch expression

This is a pretty specific case, though I did try to narrow it down as much as I could, so I'd say it's probably not the highest of priorities.
Create a project based on the given sample source file. You'll need at least Java 14 to use the switch expression syntax :

**Main.java**
```
package org.example;
public class Main {
	public static String format (String... input) {
		return "";
	}
	public static void main(String[] args) {
		int value = 0;
		String message = switch (value) {
			case 0 -> format("");
			default -> "";
		};
	}
}
```

- Select the only `format` reference in `main()`
- `Refactor -> Inline` using whichever method you prefer (eg. `Refactor` menu item, or context menu)
- Select `Preview` or `Ok` (also I think playing with any of the options shouldn't matter)
- You should see the following error in the Error Log view :

<details>
<summary>stacktrace</summary>

```
java.lang.reflect.InvocationTargetException
	at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:395)
	at org.eclipse.ltk.internal.ui.refactoring.RefactoringWizardDialog2.run(RefactoringWizardDialog2.java:324)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizard.createChange(RefactoringWizard.java:662)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizard.computeUserInputSuccessorPage(RefactoringWizard.java:504)
	at org.eclipse.ltk.ui.refactoring.UserInputWizardPage.computeSuccessorPage(UserInputWizardPage.java:77)
	at org.eclipse.ltk.ui.refactoring.UserInputWizardPage.getNextPage(UserInputWizardPage.java:113)
	at org.eclipse.ltk.internal.ui.refactoring.RefactoringWizardDialog2.nextOrPreviewPressed(RefactoringWizardDialog2.java:493)
	at org.eclipse.ltk.internal.ui.refactoring.RefactoringWizardDialog2$1.widgetSelected(RefactoringWizardDialog2.java:693)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5794)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1529)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5025)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4477)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.lambda$0(RefactoringWizardOpenOperation.java:190)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:74)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.run(RefactoringWizardOpenOperation.java:209)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.run(RefactoringWizardOpenOperation.java:126)
	at org.eclipse.jdt.internal.ui.refactoring.actions.RefactoringStarter.activate(RefactoringStarter.java:41)
	at org.eclipse.jdt.internal.corext.refactoring.RefactoringExecutionStarter.startInlineMethodRefactoring(RefactoringExecutionStarter.java:336)
	at org.eclipse.jdt.internal.ui.refactoring.actions.InlineMethodAction.tryInlineMethod(InlineMethodAction.java:151)
	at org.eclipse.jdt.ui.actions.InlineAction.run(InlineAction.java:127)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.dispatchRun(SelectionDispatchAction.java:278)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.run(SelectionDispatchAction.java:252)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5794)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1529)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5025)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4477)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.jdt.core.dom.rewrite.ListRewrite.insertAt(org.eclipse.jdt.core.dom.ASTNode, int, org.eclipse.text.edits.TextEditGroup)" because "this.fListRewrite" is null
	at org.eclipse.jdt.internal.corext.refactoring.code.CallInliner.addNewLocals(CallInliner.java:569)
	at org.eclipse.jdt.internal.corext.refactoring.code.CallInliner.perform(CallInliner.java:456)
	at org.eclipse.jdt.internal.corext.refactoring.code.InlineMethodRefactoring.checkFinalConditions(InlineMethodRefactoring.java:315)
	at org.eclipse.ltk.core.refactoring.CheckConditionsOperation.run(CheckConditionsOperation.java:86)
	at org.eclipse.ltk.core.refactoring.CreateChangeOperation.run(CreateChangeOperation.java:122)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2380)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2405)
	at org.eclipse.ltk.internal.ui.refactoring.WorkbenchRunnableAdapter.run(WorkbenchRunnableAdapter.java:89)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
```
</details>

--- a/org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/CallInliner.java
+++ b/org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/CallInliner.java
@@ -78,12 +78,14 @@ import org.eclipse.jdt.core.dom.ReturnStatement;
 import org.eclipse.jdt.core.dom.SimpleName;
 import org.eclipse.jdt.core.dom.Statement;
 import org.eclipse.jdt.core.dom.SuperFieldAccess;
+import org.eclipse.jdt.core.dom.SwitchExpression;
 import org.eclipse.jdt.core.dom.SwitchStatement;
 import org.eclipse.jdt.core.dom.ThisExpression;
 import org.eclipse.jdt.core.dom.Type;
 import org.eclipse.jdt.core.dom.VariableDeclarationFragment;
 import org.eclipse.jdt.core.dom.VariableDeclarationStatement;
 import org.eclipse.jdt.core.dom.WhileStatement;
+import org.eclipse.jdt.core.dom.YieldStatement;
 import org.eclipse.jdt.core.dom.rewrite.ASTRewrite;
 import org.eclipse.jdt.core.dom.rewrite.ImportRewrite;
 import org.eclipse.jdt.core.dom.rewrite.ImportRewrite.ImportRewriteContext;
@@ -135,6 +137,7 @@ public class CallInliner {
 	private CodeScopeBuilder.Scope fInvocationScope;
 	private boolean fFieldInitializer;
 	private List<VariableDeclarationStatement> fLocals;
+	private Block fBlock;
 	private CallContext fContext;
 
 	private class InlineEvaluator extends HierarchicalASTVisitor {
@@ -206,6 +209,7 @@ public class CallInliner {
 		fRewrite= ASTRewrite.create(targetAstRoot.getAST());
 		fRewrite.setTargetSourceRangeComputer(new NoCommentSourceRangeComputer());
 		fTypeEnvironment= new TypeEnvironment();
+		fBlock = null;
 	}
 
 	public void dispose() {
@@ -280,7 +284,9 @@ public class CallInliner {
 	private void initializeTargetNode() {
 		ASTNode parent= fInvocation.getParent();
 		int nodeType= parent.getNodeType();
-		if (nodeType == ASTNode.EXPRESSION_STATEMENT || nodeType == ASTNode.RETURN_STATEMENT) {
+		if (nodeType == ASTNode.EXPRESSION_STATEMENT
+				|| nodeType == ASTNode.RETURN_STATEMENT
+				|| nodeType == ASTNode.YIELD_STATEMENT) {
 			fTargetNode= parent;
 		} else {
 			fTargetNode= fInvocation;
@@ -562,12 +568,45 @@ public class CallInliner {
 	}
 
 	private void addNewLocals(TextEditGroup textEditGroup) {
+		fBlock = null;
 		if (fLocals.isEmpty())
 			return;
-		for (VariableDeclarationStatement variableDeclarationStatement : fLocals) {
-			ASTNode element= variableDeclarationStatement;
-			fListRewrite.insertAt(element, fInsertionIndex++, textEditGroup);
+		if (needToCreateBlockStatement()) {
+			Block block= fTargetNode.getAST().newBlock();
+
+			for (VariableDeclarationStatement variableDeclarationStatement : fLocals) {
+				ASTNode element= variableDeclarationStatement;
+				block.statements().add(element);
+			}
+			fRewrite.replace(fTargetNode, block, textEditGroup);
+			fBlock = block;
 		}
+		else {
+			for (VariableDeclarationStatement variableDeclarationStatement : fLocals) {
+				ASTNode element= variableDeclarationStatement;
+				fListRewrite.insertAt(element, fInsertionIndex++, textEditGroup);
+			}
+		}
+	}
+
+	private boolean needToCreateBlockStatement() {
+		boolean need = false;
+		if (fContext != null && ASTNode.YIELD_STATEMENT == fContext.callMode) {
+			if (fTargetNode !=  null && fTargetNode instanceof YieldStatement) {
+				ASTNode parent = fTargetNode;
+				while (parent != null) {
+					if (parent instanceof SwitchExpression) {
+						need = true;
+						break;
+					}
+					if (parent instanceof Block) {
+						break;
+					}
+					parent = parent.getParent();
+				}
+			}
+		}
+		return need;
 	}
 
 	private void replaceCall(RefactoringStatus status, String[] blocks, TextEditGroup textEditGroup) {
@@ -632,6 +671,12 @@ public class CallInliner {
 					pExp.setExpression((Expression)node);
 					node= pExp;
 				}
+			} else if (fContext.callMode == ASTNode.YIELD_STATEMENT) {
+				if (fBlock != null) {
+					node= fRewrite.createStringPlaceholder(block, ASTNode.BLOCK);
+				} else {
+					node= fRewrite.createStringPlaceholder(block, ASTNode.YIELD_STATEMENT);
+				}
 			} else {
 				node= fRewrite.createStringPlaceholder(block, ASTNode.RETURN_STATEMENT);
 			}
@@ -640,6 +685,8 @@ public class CallInliner {
 			if (node != null) {
 				if (fTargetNode == null) {
 					fListRewrite.insertAt(node, fInsertionIndex++, textEditGroup);
+				} else if (fBlock != null) {
+					fBlock.statements().add(node);
 				} else {
 					fRewrite.replace(fTargetNode, node, textEditGroup);
 				}
@@ -772,6 +819,10 @@ public class CallInliner {
 		} else if (type == ASTNode.SWITCH_STATEMENT) {
 			SwitchStatement switchStatement= (SwitchStatement)container;
 			fListRewrite= fRewrite.getListRewrite(switchStatement, SwitchStatement.STATEMENTS_PROPERTY);
+			fInsertionIndex= fListRewrite.getRewrittenList().indexOf(parentStatement);
+		} else if (type == ASTNode.SWITCH_EXPRESSION) {
+			SwitchExpression switchExpression= (SwitchExpression)container;
+			fListRewrite= fRewrite.getListRewrite(switchExpression, SwitchExpression.STATEMENTS_PROPERTY);
 			fInsertionIndex= fListRewrite.getRewrittenList().indexOf(parentStatement);
 		} else if (isControlStatement(container) || type == ASTNode.LABELED_STATEMENT) {
 			fNeedsStatement= true;

--- a/org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/SourceProvider.java
+++ b/org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/SourceProvider.java
@@ -87,6 +87,7 @@ import org.eclipse.jdt.core.dom.StructuralPropertyDescriptor;
 import org.eclipse.jdt.core.dom.SwitchCase;
 import org.eclipse.jdt.core.dom.ThisExpression;
 import org.eclipse.jdt.core.dom.WhileStatement;
+import org.eclipse.jdt.core.dom.YieldStatement;
 import org.eclipse.jdt.core.dom.rewrite.ASTRewrite;
 import org.eclipse.jdt.core.dom.rewrite.ImportRewrite;
 import org.eclipse.jdt.core.dom.rewrite.ImportRewrite.ImportRewriteContext;
@@ -370,10 +371,11 @@ public class SourceProvider {
 		updateStaticReferences(rewriter, context);
 		updateTypeVariables(rewriter, context);
 		updateMethodTypeVariable(rewriter, context);
-
+		updateReturnStatements(rewriter, context);
 		List<IRegion> ranges= null;
 		if (hasReturnValue()) {
-			if (context.callMode == ASTNode.RETURN_STATEMENT) {
+			if (context.callMode == ASTNode.RETURN_STATEMENT
+					|| context.callMode == ASTNode.YIELD_STATEMENT) {
 				ranges= getStatementRanges();
 			} else {
 				ranges= getExpressionRanges();
@@ -619,6 +621,23 @@ public class SourceProvider {
 		if (method == null)
 			return;
 		rewriteReferences(rewriter, method.getTypeArguments(), fAnalyzer.getMethodTypeParameterReferences());
+	}
+
+	private void updateReturnStatements(ASTRewrite rewriter, CallContext context) {
+		if (rewriter != null && context != null && context.callMode == ASTNode.YIELD_STATEMENT) {
+			Block nodeToVisit= fDeclaration.getBody();
+			ASTVisitor visitor= new ASTVisitor() {
+				@Override
+				public boolean visit(ReturnStatement node) {
+					Expression exp= node.getExpression();
+					YieldStatement yStmt= rewriter.getAST().newYieldStatement();
+					yStmt.setExpression((Expression) rewriter.createMoveTarget(exp));
+					rewriter.replace(node, yStmt, null);
+					return false;
+				}
+			};
+			nodeToVisit.accept(visitor);
+		}
 	}
 
 	private void rewriteReferences(ASTRewrite rewriter, ITypeBinding[] typeArguments, List<NameData> typeParameterReferences) {

--- a/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_in/TestSwitchExpression1.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_in/TestSwitchExpression1.java
@@ -1,0 +1,16 @@
+package simple18_in;
+
+import java.lang.String;
+
+public class TestSwitchExpression1 {
+	public static String format (String... input) {
+		return "";
+	}
+	public static void main(String[] args) {
+		int value = 0;
+		String message = switch (value) {
+			case 0 -> /*]*/format("")/*[*/;
+			default -> "";
+		};
+	}
+}

--- a/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_in/TestSwitchExpression2.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_in/TestSwitchExpression2.java
@@ -1,0 +1,16 @@
+package simple18_in;
+
+import java.lang.String;
+
+public class TestSwitchExpression1 {
+	public static String format (String... input) {
+		return "";
+	}
+	public static void main(String[] args) {
+		int value = 0;
+		String message = switch (value) {
+			case 0 : yield /*]*/format("")/*[*/;
+			default : yield "";
+		};
+	}
+}

--- a/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_in/TestSwitchExpression3.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_in/TestSwitchExpression3.java
@@ -1,0 +1,22 @@
+package simple18_in;
+
+import java.lang.String;
+
+public class TestSwitchExpression1 {
+	public static String format (String... input) {
+		if (input == null) {
+			return "null";
+		} else if (input.length > 1) {
+			return "multiple";
+		} else {
+			return "single";
+		}
+	}
+	public static void main(String[] args) {
+		int value = 0;
+		String message = switch (value) {
+			case 0 -> /*]*/format("")/*[*/;
+			default -> "";
+		};
+	}
+}

--- a/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_in/TestSwitchExpression4.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_in/TestSwitchExpression4.java
@@ -1,0 +1,22 @@
+package simple18_in;
+
+import java.lang.String;
+
+public class TestSwitchExpression1 {
+	public static String format (String... input) {
+		if (input == null) {
+			return "null";
+		} else if (input.length > 1) {
+			return "multiple";
+		} else {
+			return "single";
+		}
+	}
+	public static void main(String[] args) {
+		int value = 0;
+		String message = switch (value) {
+			case 0 : yield /*]*/format("")/*[*/;
+			default : yield "";
+		};
+	}
+}

--- a/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_out/TestSwitchExpression1.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_out/TestSwitchExpression1.java
@@ -1,0 +1,19 @@
+package simple18_in;
+
+import java.lang.String;
+
+public class TestSwitchExpression1 {
+	public static String format (String... input) {
+		return "";
+	}
+	public static void main(String[] args) {
+		int value = 0;
+		String message = switch (value) {
+			case 0 -> {
+	String[] input = {""};
+	yield "";
+}
+			default -> "";
+		};
+	}
+}

--- a/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_out/TestSwitchExpression2.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_out/TestSwitchExpression2.java
@@ -1,0 +1,19 @@
+package simple18_in;
+
+import java.lang.String;
+
+public class TestSwitchExpression1 {
+	public static String format (String... input) {
+		return "";
+	}
+	public static void main(String[] args) {
+		int value = 0;
+		String message = switch (value) {
+			case 0 : {
+					String[] input = {""};
+					yield "";
+				}
+			default : yield "";
+		};
+	}
+}

--- a/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_out/TestSwitchExpression3.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_out/TestSwitchExpression3.java
@@ -1,0 +1,31 @@
+package simple18_in;
+
+import java.lang.String;
+
+public class TestSwitchExpression1 {
+	public static String format (String... input) {
+		if (input == null) {
+			return "null";
+		} else if (input.length > 1) {
+			return "multiple";
+		} else {
+			return "single";
+		}
+	}
+	public static void main(String[] args) {
+		int value = 0;
+		String message = switch (value) {
+			case 0 -> {
+	String[] input = {""};
+	if (input == null) {
+		yield "null";
+	} else if (input.length > 1) {
+		yield "multiple";
+	} else {
+		yield "single";
+	}
+}
+			default -> "";
+		};
+	}
+}

--- a/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_out/TestSwitchExpression4.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/simple14_out/TestSwitchExpression4.java
@@ -1,0 +1,31 @@
+package simple18_in;
+
+import java.lang.String;
+
+public class TestSwitchExpression1 {
+	public static String format (String... input) {
+		if (input == null) {
+			return "null";
+		} else if (input.length > 1) {
+			return "multiple";
+		} else {
+			return "single";
+		}
+	}
+	public static void main(String[] args) {
+		int value = 0;
+		String message = switch (value) {
+			case 0 : {
+					String[] input = {""};
+					if (input == null) {
+						yield "null";
+					} else if (input.length > 1) {
+						yield "multiple";
+					} else {
+						yield "single";
+					}
+				}
+			default : yield "";
+		};
+	}
+}

--- a/org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/InlineMethodTests16.java
+++ b/org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/InlineMethodTests16.java
@@ -64,4 +64,24 @@ public class InlineMethodTests16 extends InlineMethodTests {
 	public void testInlineProblem2() throws Exception {
 		performDefaultTest();
 	}
+
+	@Test
+	public void testSwitchExpression1() throws Exception {
+		performSimpleTest();
+	}
+
+	@Test
+	public void testSwitchExpression2() throws Exception {
+		performSimpleTest();
+	}
+
+	@Test
+	public void testSwitchExpression3() throws Exception {
+		performSimpleTest();
+	}
+
+	@Test
+	public void testSwitchExpression4() throws Exception {
+		performSimpleTest();
+	}
 }

